### PR TITLE
CP-15502: Adding VM Scheduled Snapshot functions, APIs, CLIs and records.

### PR DESCRIPTION
**Pull request by sharady**

### CODE REVIEW

File: ocaml/client_records/record_util.ml [hunk 1]
```diff
@@ -156,6 +156,28 @@ let string_to_vif_locking_mode = function
...
+let vmss_type_to_string = function
+	| `snapshot -> "snapshot"
+	| `checkpoint -> "checkpoint"
+	| `snapshot_with_quiesce -> "snapshot_with_quiesce"
+
+let string_to_vmss_type = function
+	| "snapshot" -> `snapshot
+	| "checkpoint" -> `checkpoint
+	| "snapshot_with_quiesce" -> `snapshot_with_quiesce
+	| _ -> raise (Record_failure ("Expected 'snapshot', 'checkpoint', 'snapshot_with_quiesce'"))
```
Comment (euanh) [inline ocaml/client_records/record_util.ml]: Can you include the string which was actually received in this exception, as `string_to_network_default_locking mode` does?

File: ocaml/client_records/record_util.ml [hunk 2]
```diff
@@ -156,6 +156,28 @@ let string_to_vif_locking_mode = function
...
+let vmss_frequency_to_string = function
+	| `hourly -> "hourly"
+	| `daily -> "daily"
+	| `weekly -> "weekly"
+
+let string_to_vmss_frequency = function
+	| "hourly" -> `hourly
+	| "daily" -> `daily
+	| "weekly" -> `weekly
+	| _ -> raise (Record_failure ("Expected 'hourly', 'daily', 'weekly'"))
```
Comment (euanh) [inline ocaml/client_records/record_util.ml]: Same comment as on `string_to_vmss_type`

**Comment by euanh** [comment]
I notice that the design is called 'scheduled snapshots' but in the rest of the design document and in the code here (both external strings and internal definitions) you use 'schedule snapshot'.   'Scheduled snapshot' seems more correct to me.

File: ocaml/xapi/xapi_vmss.ml [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+				(* add missing keys with default values *)
+				let value = List.map (fun (k,(kt,default))->if List.mem_assoc k value then (k,(List.assoc k value)) else (k,default)) ks in
+				
+				(* remove extra unexpected keys *)
+				let value = List.fold_right (fun (k,v) acc->if List.mem_assoc k ks then (k,v)::acc else acc) value [] in
+				(* for this ks, each key value must be valid *)
+				List.map (fun (k,v)-> k,(assert_key ~field ~ks ~key:k ~value:v)) value
+			)
+			else None
+		)
```
Comment (euanh) [inline ocaml/xapi/xapi_vmss.ml]: All the functions from `mem` above to here seem like generic utilities which should be in a general library module.
They should all have unit tests.

**Comment by euanh** [comment]
Does the actual snapshot scheduler apply any jitter to the schedule?   There must be some mechanism to spread out the load of taking the snapshots - otherwise in the worst case every 15 minutes Xapi will grind to a halt as it snapshots every VM at the same time.

File: ocaml/xapi/xapi_vmss.ml (hunk 3 above)
Comment (johnelse) [inline ocaml/xapi/xapi_vmss.ml]: There are already some functions like this for handling key-value pair lists in map_check.ml - that might be a good place to add them.

**Comment by sharady** [comment]
@euanh @johnelse  Thanks for the review :)
I have updated the cosmetic changes. Few more coming soon.

File: ocaml/xapi/xapi_vmss.ml [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+
+let set_schedule ~__context ~self ~value =
+	let value = assert_keys ~ty:"" ~ks:schedule_all_keys ~value ~db:(Db.VMSS.get_schedule ~__context ~self) in
+	Db.VMSS.set_schedule ~__context ~self ~value
+
+let set_type ~__context ~self ~value =
+	(* Check VMs associated to VMSS supports the new snapshot type *)
+	let snapshot_type = Record_util.vmss_type_to_string value in
+	let vms = Db.VMSS.get_VMs ~__context ~self in
+	if not (vms = []) then
```
Comment (euanh) [inline ocaml/xapi/xapi_vmss.ml]: if vms <> [] then ...

**Comment by simonjbeaumont** [comment]
Removing the reviewed label since there are more changes coming according to sharad's comment.

**Comment by sharady** [comment]
@euanh I have moved the key:value check functions to map_check.ml
Please have a look on the PR. thanks

**Comment by simonjbeaumont** [comment]
@sharady you can remove the needs updating label if you think it's ready for (re-)review.

**Comment by stephen-turner** [comment]
I haven't reviewed this PR, but I did notice some terminology problems.
A VMSS is a "VM Snapshot Schedule" not a "VM Scheduled Snapshot".
A scheduled snapshot is a single snapshot, created according to the schedule.
This affects some function names as well as comments.
For example, vm_set_scheduled_snapshot should be vm_set_snapshot_schedule.

**Comment by jonludlam** [comment]
closing this until we expect progress

**Pull request closed without merging**